Update RawMessageCodec and tests

alextaka opened the pull request:
Adding ability to decode TransferAccessResponse and adding associated test skeletons

alextaka committed:
Update RawMessageCodec and tests

Adding ability to decode TransferAccessResponse and adding associated test skeletons

alextaka commented:
@leshi PTAL

leshi commented on u2f-ref-code/java/src/com/google/u2f/codec/RawMessageCodec.java:
```diff
@@ -192,6 +194,80 @@ public static AuthenticateResponse decodeAuthenticateResponse(byte[] data) throw
 
+  public static TransferAccessResponse decodeTransferAccessResponse(byte[] data) 
+      throws U2FException {
+    final int SEQUENCE_NUMBER_SIZE = 1;
```
while it's fine, you don't actually need the final qualifier here

alextaka replied:
fixed to make them standard local variables.

leshi replied:
I don't see no changes -- did you forget to push?

alextaka replied:
pushed.

leshi commented on u2f-ref-code/java/src/com/google/u2f/codec/RawMessageCodec.java:
```diff
@@ -192,6 +194,80 @@ public static AuthenticateResponse decodeAuthenticateResponse(byte[] data) throw
+	  DataInputStream inputStream = new DataInputStream(new ByteArrayInputStream(data));
+	  byte controlByte = inputStream.readByte();
+	  
+	  if (!inputStream.markSupported()) {
```
why would it not support mark?

alextaka replied:
I'm not actually sure. I just wanted to make sure that if this is the reason the program failed that it was clear. Is this a bad reason to have the check?

alextaka replied:
pinging this question, @leshi

leshi commented on u2f-ref-code/java/src/com/google/u2f/codec/RawMessageCodec.java:
```diff
@@ -192,6 +194,80 @@ public static AuthenticateResponse decodeAuthenticateResponse(byte[] data) throw
+	    throw new U2FException("Stream doesn't support mark and reset.");
+	  }
+	  inputStream.mark(inputStream.available());
+      int numberOfTransferAccessMessages = inputStream.readUnsignedByte();
```
indentation looks off

alextaka replied:
odd, it wasn't like that on my copy. Hopefully fixed. We'll see on the next submission.

leshi commented on u2f-ref-code/java/src/com/google/u2f/codec/RawMessageCodec.java:
```diff
@@ -192,6 +194,80 @@ public static AuthenticateResponse decodeAuthenticateResponse(byte[] data) throw
+        inputStream.reset();
+
+        int lengthOfRawTransferAccessMessage =
+            SEQUENCE_NUMBER_SIZE + PUBLIC_KEY_SIZE + APP_ID_SIZE + 1 + lengthOfAttestationCertificate
```
what's the +1?  (make a comment somewhere)

alextaka replied:
The +1 is the length byte. I put the first one in the wrong place. I can make it it's own variable (lengthField_signatureUsingAttestationKey, lengthField_signatureUsingAuthenticationKey), or just put a comment in there that the +1's are for the length bytes for signatures.

alextaka replied:
Currently have made it its own variable.

leshi commented on u2f-ref-code/java/src/com/google/u2f/codec/RawMessageCodec.java:
```diff
@@ -192,6 +194,80 @@ public static AuthenticateResponse decodeAuthenticateResponse(byte[] data) throw
+            TransferAccessMessage.fromBytes(rawTransferAccessMessage);
+        if (transferAccessMessages[numberOfTransferAccessMessages - i - 1]
+            .getMessageSequenceNumber() != numberOfTransferAccessMessages - i) {
+          throw new U2FException("Messages not in order, unexpected sequence number");
```
For debugging, print the expected sequence number and the actual one.

alextaka replied:
done.

leshi commented on u2f-ref-code/java/src/com/google/u2f/codec/RawMessageCodec.java:
```diff
@@ -215,4 +291,65 @@ public static AuthenticateResponse decodeAuthenticateResponse(byte[] data) throw
+    byte[] signedData = new byte[1 + newUserPublicKey.length + applicationSha256.length
+        + attestationCertificateBytes.length];
+    
+    ByteBuffer.wrap(signedData)
```
I would just use https://docs.oracle.com/javase/7/docs/api/java/io/ByteArrayOutputStream.html

Something like:

```
return new ByteArraOutputStream()
    .put(sequenceNumber)
    .put(newUserPublicKey)
    .put(applicationSha256)
    .put(attestationCertificateBytes)
    .put(signatureUsingAuthenticationKey)
    .toByteArray();
```

alextaka replied:
I used byte buffer because that's what is used elsewhere in the code. It also helped me catch a couple missing fields (for example the two extra '+1's for length fields you pointed out in another comment)  but this may not be a good reason to use it. What is the reason that the output stream is better?

alextaka replied:
Also, I believe this would actually have to look like this:

```
ByteArrayOutputStream signedDataOutputStream = new ByteArrayOutputStream();
signedDataOutputStream.write(sequenceNumber);
signedDataOutputStream.write(newUserPublicKey);
...
signedDataOutputStream.write(signatureUsingAuthenticationKey)

return signedDataOutputStream.toByteArray();
```

This is because there is no "put()" method, just write(), which returns void.

alextaka replied:
I have currently left it as is.

leshi replied:
It's just hard to make sure the size is correct when you pre-allocate.  It's also hard to maintain.  I strongly suggest not pre-allocating.

alextaka replied:
done. changed. Do you want me to update the other functions (written by someone else) to ByteArrayOutputStream in another PR? or is that too much? It is all over the codebase.

leshi commented on u2f-ref-code/java/src/com/google/u2f/codec/RawMessageCodec.java:
```diff
@@ -215,4 +291,65 @@ public static AuthenticateResponse decodeAuthenticateResponse(byte[] data) throw
+    byte[] signedData = new byte[1 + newUserPublicKey.length + applicationSha256.length
+        + attestationCertificateBytes.length + signatureUsingAuthenticationKey.length];
+    
+    ByteBuffer.wrap(signedData)
```
same as above

alextaka replied:
done

leshi commented on u2f-ref-code/java/src/com/google/u2f/codec/RawMessageCodec.java:
```diff
@@ -215,4 +291,65 @@ public static AuthenticateResponse decodeAuthenticateResponse(byte[] data) throw
+    byte[] signedData = new byte[1 + 4 + challengeSha256.length + keyHandle.length
+        + lastTransferAccessMessageSha256.length];
+
+    ByteBuffer.wrap(signedData)
```
ditto

alextaka replied:
done

leshi commented on u2f-ref-code/java/tests/com/google/u2f/codec/RawCodecTest.java:
```diff
@@ -108,4 +116,99 @@ public void testEncodeAuthenticateSignedBytes() throws Exception {
+  
+  @Test
+  public void testDecodeTransferAccessResponse_extraBytes() throws Exception {
+    // This should just be the same as an invalid signature
```
tests need some love.

alextaka replied:
they're filled out. done.

alextaka commented on u2f-ref-code/java/src/com/google/u2f/codec/RawMessageCodec.java:
```diff
@@ -192,6 +194,80 @@ public static AuthenticateResponse decodeAuthenticateResponse(byte[] data) throw
+            TransferAccessMessage.fromBytes(rawTransferAccessMessage);
+        if (transferAccessMessages[numberOfTransferAccessMessages - i - 1]
+            .getMessageSequenceNumber() != numberOfTransferAccessMessages - i) {
+          throw new U2FException("Messages not in order, unexpected sequence number");
```
The first thing that comes to mind is to use a string builder and formatter to build the string, and then call 
throw new U2fException(builtStringBuilder.toString());

does this seem acceptable?

alextaka commented on u2f-ref-code/java/src/com/google/u2f/codec/RawMessageCodec.java:
```diff
@@ -192,6 +194,80 @@ public static AuthenticateResponse decodeAuthenticateResponse(byte[] data) throw
+      byte[] signature = new byte[inputStream.available()];
+	  inputStream.readFully(signature);
+	  
+	  if (inputStream.available() != 0) {
```
Do we need this check? It should be impossible to get here.

alextaka replied:
I've removed it for now.

alextaka committed:
Addressed first round of comments.

Haven't updated ByteArrayOutputStream, pending discussion.
Also filled in the rest of the tests

alextaka commented on u2f-ref-code/java/src/com/google/u2f/codec/RawMessageCodec.java:
```diff
@@ -215,4 +291,65 @@ public static AuthenticateResponse decodeAuthenticateResponse(byte[] data) throw
+    byte[] signedData = new byte[1 + newUserPublicKey.length + applicationSha256.length
+        + attestationCertificateBytes.length + signatureUsingAuthenticationKey.length];
+    
+    ByteBuffer.wrap(signedData)
```
see above.

alextaka commented on u2f-ref-code/java/src/com/google/u2f/codec/RawMessageCodec.java:
```diff
@@ -215,4 +291,65 @@ public static AuthenticateResponse decodeAuthenticateResponse(byte[] data) throw
+    byte[] signedData = new byte[1 + 4 + challengeSha256.length + keyHandle.length
+        + lastTransferAccessMessageSha256.length];
+
+    ByteBuffer.wrap(signedData)
```
see above

leshi commented on u2f-ref-code/java/src/com/google/u2f/codec/RawMessageCodec.java:
```diff
@@ -192,6 +194,85 @@ public static AuthenticateResponse decodeAuthenticateResponse(byte[] data) throw
+        int lengthOfSignatureUsingAttestationKey = inputStream.readUnsignedByte();
+        inputStream.reset();
+
+        int lengthField_signatureUsingAttestationKey = 1;
```
don't mix camelcase and underscores

alextaka replied:
done

leshi commented on u2f-ref-code/java/src/com/google/u2f/codec/RawMessageCodec.java:
```diff
@@ -192,6 +194,85 @@ public static AuthenticateResponse decodeAuthenticateResponse(byte[] data) throw
+
+        byte[] rawTransferAccessMessage = new byte[lengthOfRawTransferAccessMessage];
+        inputStream.readFully(rawTransferAccessMessage);
+        transferAccessMessages[numberOfTransferAccessMessages - i - 1] =
```
things like -1 make me nervous.  add some comments to explain what you're doing.

alextaka replied:
done. created a variable to make things more clear.

leshi commented on u2f-ref-code/java/src/com/google/u2f/codec/RawMessageCodec.java:
```diff
@@ -192,6 +194,85 @@ public static AuthenticateResponse decodeAuthenticateResponse(byte[] data) throw
+      TransferAccessMessage[] transferAccessMessages =
+          new TransferAccessMessage[numberOfTransferAccessMessages];
+
+      for (int i = 0; i < numberOfTransferAccessMessages; i++) {
```
add some comments to explain what this loop is doing.

leshi replied:
the body of this for loop is almost asking to be its own method.  Try it and see if it works.

alextaka replied:
Done, but a potential issue may need to be discussed:

If I pass a [DataInputStream](https://docs.oracle.com/javase/7/docs/api/java/io/DataInputStream.html), I don't know that there is a way to check the underlying stream to make sure it is at least the type we expect ([ByteArrayInputStream](https://docs.oracle.com/javase/7/docs/api/java/io/ByteArrayInputStream.html)). I can, however, just treat it as a byte array since this should be the only function calling it for now.

leshi commented on u2f-ref-code/java/tests/com/google/u2f/codec/RawCodecTest.java:
```diff
@@ -6,19 +6,28 @@
 public class RawCodecTest extends TestVectors {
+  private static final int INITIAL_COUNTER = 0;
+  private final Crypto crypto = new BouncyCastleCrypto();
+
```
extra new line

alextaka replied:
done.

alextaka committed:
Addressed second round of comments.

leshi commented on u2f-ref-code/java/src/com/google/u2f/codec/RawMessageCodec.java:
```diff
@@ -192,6 +195,113 @@ public static AuthenticateResponse decodeAuthenticateResponse(byte[] data) throw
+      byte controlByte = inputStream.readByte();
+
+      if (!inputStream.markSupported()) {
+        throw new U2FException("Stream doesn't support mark and reset.");
```
I don't think you need to test for it.  My hunch is the code below will throw if there is no support (you'll catch it your catch statement and will re-throw)

alextaka replied:
ok, thanks. removed.

leshi commented on u2f-ref-code/java/src/com/google/u2f/codec/RawMessageCodec.java:
```diff
@@ -192,6 +195,113 @@ public static AuthenticateResponse decodeAuthenticateResponse(byte[] data) throw
+            + lengthOfSignatureUsingPrivateKey + sizeOfLengthFieldForSignatureUsingAttestationKey
+            + lengthOfSignatureUsingAttestationKey;
+
+        byte[] rawTransferAccessMessage = new byte[lengthOfRawTransferAccessMessage];
```
use an output stream here as well.

alextaka replied:
This is an input stream, not an output stream. I think we may need this to tell it how many bytes to read as it is just reading in the whole thing, not writing it into the buffer in pieces.

leshi commented on u2f-ref-code/java/src/com/google/u2f/codec/RawMessageCodec.java:
```diff
@@ -215,4 +325,71 @@ public static AuthenticateResponse decodeAuthenticateResponse(byte[] data) throw
+    
+    ByteArrayOutputStream signedDataOutputStream = new ByteArrayOutputStream();
+    try {
+      signedDataOutputStream.write(sequenceNumber);
```
so much nicer!

alextaka commented on u2f-ref-code/java/src/com/google/u2f/codec/RawMessageCodec.java:
```diff
@@ -192,6 +195,113 @@ public static AuthenticateResponse decodeAuthenticateResponse(byte[] data) throw
+      byte controlByte = inputStream.readByte();
+
+      if (!inputStream.markSupported()) {
+        throw new U2FException("Stream doesn't support mark and reset.");
```
ok. thanks. removed.

alextaka committed:
Addressed third round of commits, except the input/output stream
